posts/loggingPython.md

```diff
@@ -0,0 +1,325 @@
+
+Some of this data, such as `time` and `levelname` can be captured automatically, but if you can (and should) push `extra` context to your logs.
+
+```python
+logger = logging.getLogger(__name__)
+logger.info('purchase completed', extra={‘user': 'Sid Panjwani'})
+```
+
+If you’re wondering why you do not see your logs, remember to change the `level`. By default, it’s set to `warning`.
+
+*Yep, it’s that’s easy…*
+
+### Tracking Exceptions through Logs
+
+Some companies have made it their core competency helping you capture and react to exceptions. If you’re just a hobbyist, it’s pretty easy to capture your errors yourself and write them to your logs.
+
+```python
+def captureException():
+    # this should do something
+
+try:
+    1/0
+except:
+    captureException()
+```
+
+When capturing an exception, it’s essential to add context. Similar to when you’re debugging, you want to know who the user is and what they were doing when the exception struck.
+
+```python
+import sys
+
+# remember to set up your logger
+
+def captureException():
+    logger.warning(sys.exc_info())
+```
+
+![](./images/loggingPython/exception1.png)
+
+This gives us a `traceback` object that we can use to see the stack from our exception. We didn’t even need to pass anything into our `captureException` method.
+
+Let’s figure out what the `traceback` object gives us and clean up this information a little.
+
+```python
+import sys
+import traceback
+
+def captureException():
+    r = list(sys.exc_info())
+
+    e = dict()
+    e["name"] = r[1]
+```
+
+![](./images/loggingPython/exception2.png)
+
+Now, this can be cleaned up using some string manipulation in Python.
+
+## Storing & Accessing These Logs
+
+Now that you've learned to write these (beautiful) logs, you've got to learn what to do with them. I mean it would be a waste to write them to the console and leave them to be deleted when the program closes.
+
+### Logging to a File
+
+Generally, you don't (just) want to log to the console for them to be deleted immediately. The logging module makes it easy to write your logs to a local file using a `handler`.
+
+```python
+import logging
+logger = logging.getLogger(__name__)
+
+handler = logging.FileHandler('myLogs.log')
+handler.setLevel(logging.INFO)
+
+logger.addHandler(handler)
+logger.info('You can find this written in myLogs.log')
+```
+
+### Logging to the Cloud
+
+Writing your logs to the cloud seriously makes your life easier, it serves as a layer of abstraction so you don’t have to worry how they are getting to the service and you can focus your time on what’s important.
```

Comment:
It would be better to refer to Timber as a "hosted log aggregation service" since "the cloud" is a vague term. Also, the clause "it serves as a layer of abstraction so you don’t have to worry how they are getting to the service" does not make sense.